crates/pallet-rewards/src/lib.rs

```diff
@@ -97,40 +188,265 @@ mod pallet {
+            .max_component_issuance
+            .checked_sub(&total_initial_subsidy);
+        let Some(max_decay_issuance) = maybe_max_decay_issuance else {
+            warn!(
+                "`max_decay_issuance` underflow when calculating reference subsidy, incorrect \
+                issuance parameters"
+            );
 
-        for voter in T::FindVotingRewardAddresses::find_voting_reward_addresses() {
-            let _imbalance = T::Currency::deposit_creating(&voter, reward);
-            T::OnReward::on_reward(voter.clone(), reward);
+            return Zero::zero();
+        };
 
-            Self::deposit_event(Event::VoteReward { voter, reward });
+        // Code below is the same as:
+        // ```
+        // k = params.initial_subsidy / max_decay_issuance
+        // params.initial_subsidy*(1 - k*t + (k*t)^2/2 - (k*t)^3/6 + (k*t)^4/24 - (k*t)^5/120)
+        // ```
+        // But written with just integers.
+        //
+        // Step 1 (multiply everything by `initial_subsidy`):
+        // ```
+        // is = params.initial_subsidy
+        // mdi = max_decay_issuance
+        // is
+        //   - is*is/mdi*t
+        //   + is*(is/mdi*t)^2/2
+        //   - is*(is/mdi*t)^3/6
+        //   + is*(is/mdi*t)^4/24
+        //   - is*(is/mdi*t)^5/120
+        // ```
+        // Step 2 (expand powers):
+        // ```
+        // is
+        //   - is*is/mdi*t
+        //   + is*is*is/mdi/mdi*t*t/2
+        //   - is*is*is*is/mdi/mdi/mdi*t*t*t/6
+        //   + is*is*is*is*is/mdi/mdi/mdi/mdi*t*t*t*t/24
+        //   - is*is*is*is*is*is/mdi/mdi/mdi/mdi/mdi*t*t*t*t*t/120
+        // ```
+        // Step 3 (reorder to avoid integer overflow or zeroing):
+        // ```
+        // is
+        //   - is*is*t/mdi
+        //   + is*is*t/mdi*is*t/mdi/2
+        //   - is*is*t/mdi*is*t/mdi*is*t/mdi/6
+        //   + is*is*t/mdi*is*t/mdi*is*t/mdi*is*t/mdi/24
+        //   - is*is*t/mdi*is*t/mdi*is*t/mdi*is*t/mdi*is*t/mdi/120
+        // ```
+
+        let is = params.initial_subsidy;
+        let mdi = max_decay_issuance;
+
+        let maybe_subsidy = try {
+            is.checked_sub(&(is * is * t / mdi))?
+                .checked_add(&(is * is * t / mdi * is * t / mdi / 2_u32.into()))?
+                .checked_sub(&(is * is * t / mdi * is * t / mdi * is * t / mdi / 6_u32.into()))?
+                .checked_add(
+                    &(is * is * t / mdi * is * t / mdi * is * t / mdi * is * t
+                        / mdi
+                        / 24_u32.into()),
+                )?
+                .checked_sub(
+                    &(is * is * t / mdi * is * t / mdi * is * t / mdi * is * t / mdi * is * t
+                        / mdi
+                        / 120_u32.into()),
+                )?
+        };
+
+        match maybe_subsidy {
+            Some(subsidy) => subsidy,
+            None => {
+                warn!(
+                    "Underflow or overflow when calculating reference subsidy, incorrect \
```

Comment:
Just to note, underflow may also happen because of the approximation used. A true exponential function can't go below zero, but the Taylor series we used can and will after 30-50 years depending on initial subsidy we choose.

Reply:
Should this be a warning then or just a debug message in your opinion?

Reply:
I'd leave it as warn for now as it may still cover other edge cases. May be worth it to add a comment that the current approximation is expected to eventually underflow even with correct subsidy parameters